projects/ngx-charts-on-fhir/src/lib/data-layer-list/annotation-options/annotation-options.component.spec.ts

```diff
@@ -1,96 +1,76 @@
-  writeValue(obj: any): void {
-    /* intentionally empty stub */
-  }
-  registerOnChange(fn: any): void {
-    /* intentionally empty stub */
-  }
-  registerOnTouched(fn: any): void {
-    /* intentionally empty stub */
-  }
-  setDisabledState?(isDisabled: boolean): void {
-    /* intentionally empty stub */
-  }
-}
+import { MatInputHarness } from '@angular/material/input/testing';
+import { NoopAnimationsModule } from '@angular/platform-browser/animations';
+import { MatInputModule } from '@angular/material/input';
+import { MatFormFieldModule } from '@angular/material/form-field';
+import { ColorPickerModule } from '../../color-picker/color-picker.module';
 
 describe('AnnotationOptionsComponent', () => {
   let component: AnnotationOptionsComponent;
   let fixture: ComponentFixture<AnnotationOptionsComponent>;
   let colorService: DataLayerColorService;
-  let palette: string[];
+  let palette: string[] = ['#FFFFFF', '#121212', '#000000'];
+  let loader: HarnessLoader;
+
   beforeEach(async () => {
     colorService = new DataLayerColorService(palette);
     await TestBed.configureTestingModule({
+      imports: [ReactiveFormsModule, NoopAnimationsModule, MatInputModule, MatFormFieldModule, ColorPickerModule],
       declarations: [AnnotationOptionsComponent],
-      providers: [FormBuilder, { provide: DataLayerColorService, useValue: colorService }],
+      providers: [FormBuilder, { provide: DataLayerColorService, useValue: colorService }, { provide: COLOR_PALETTE, useValue: palette }],
     }).compileComponents();
 
     fixture = TestBed.createComponent(AnnotationOptionsComponent);
     component = fixture.componentInstance;
     fixture.detectChanges();
+    loader = TestbedHarnessEnvironment.loader(fixture);
   });
-  function updateForm(annotationOptions: any) {
-    component.form.controls['color'].setValue(annotationOptions.color);
-    component.form.controls['label'].setValue(annotationOptions.label);
-    component.form.controls['yMax'].setValue(annotationOptions.yMax);
-    component.form.controls['yMin'].setValue(annotationOptions.yMin);
-  }
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
 
-  it('form value should update from form changes', fakeAsync(() => {
-    let annotationOptions = { color: 'color', label: 'label', yMax: 80, yMin: 120 };
-    updateForm(annotationOptions);
-    fixture.detectChanges();
-    fixture.whenStable().then(() => {
-      expect(component.form.value).toEqual(annotationOptions);
-    });
-  }));
+  it('should update annotation label when input is changed', async () => {
+    let labelInputHarness = await loader.getHarness(MatInputHarness.with({ selector: "[id='annotationLabel']" }));
+    let labelInputControl = fixture.componentInstance.form.controls['label'];
+    const testLabel = 'test label';
+    await labelInputHarness.setValue(testLabel);
+    expect(labelInputControl.value).toBe(testLabel);
+  });
```

Comment:
This is an improvement because it is now testing that the template is wired up to the reactive form.

But it is still not testing that the model gets updated correctly.

`component.form`, although it is a public property, is not really part of the component's public API. That's just an implementation detail. Instead, the test should check the value emitted by `annotationChange` event.

Reply:
```suggestion
  it('should update annotation label when input is changed', async () => {
    let emitted: any = null;
    component.annotation = { label: { content: 'Old' } };
    component.annotationsChange.subscribe(e => emitted = e)
    const labelInputHarness = await loader.getHarness(MatInputHarness.with({ selector: "[id='annotationLabel']" }));
    await labelInputHarness.setValue('New');
    expect(emitted.label.content).toEqual('New');
  });
```

Reply:
this test is still checking the FormControl value instead of the annotationChange output